### PR TITLE
Improve initialising of JS modules for project

### PR DESCRIPTION
**Pull request by Guntrisoft**
- now use the concept of data-module attribute to trigger the
  initalising of a JS module
- Ensure all the JS is in assets/javascripts folder
- Created multi-checkbox module which can be potentially reuseable
- Created sortable-list module which uses ListJs to sort/filter
  a list of 'things' - used on guiders list

### CODE REVIEW

File: app/assets/javascripts/multi-checkbox.es6 [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    bindEvents() {
+      this.itemCheckboxes.on('change', this.handleItemChange.bind(this));
+      this.allCheckbox.on('change', this.handleAllChange.bind(this));
+    }
+
+    handleItemChange(event) {
+      const $el = $(event.currentTarget),
+      checked = this.getCheckedStatus($el);
```
Comment (benbarnett) [inline app/assets/javascripts/multi-checkbox.es6]: If we're going with the `,` approach for definitions, could we indent the lines that follow it? So

```es6
const $el = $(event.currentTarget),
  checked = this.getCheckedStatus($el);
```

File: app/assets/javascripts/multi-checkbox.es6 [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    addRemoveClass($el, checked) {
+      $el.parents('[data-multi-checkbox-group]')[checked ? 'addClass' : 'removeClass'](this.config.selectedClassName);
+    }
+
+    areAllItemsChecked() {
+      return $.grep(this.itemCheckboxes, (el) => {
+        return !this.getCheckedStatus($(el));
+      }).length === 0;
```
Comment (benbarnett) [inline app/assets/javascripts/multi-checkbox.es6]: :eyes: 👍

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -13,20 +13,20 @@
...
 <% content_for :body_end do %>
   <script>
-  var loadedCalendars = [];
+  var loadedModules = [];
   $(document).on('turbolinks:load', function() {
-    var calendars = $('[data-calendar]');
+    var modules = $('[data-module]');
 
-    if (loadedCalendars.length) {
+    if (loadedModules.length) {
```
Comment (benbarnett) [inline app/views/layouts/application.html.erb]: I suppose strictly we should check `loadedModules` for the presence of the specific module we're about to kick off, but as they're all on the same page I don't think this is a big issue for now.

File: app/assets/javascripts/multi-checkbox.es6 (hunk 1 above)
Comment (Guntrisoft) [inline app/assets/javascripts/multi-checkbox.es6]: sure

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (Guntrisoft) [inline app/views/layouts/application.html.erb]: as discussed in person, we won't worry about this for now.

File: app/views/users/index.html.erb [hunk 4]
```diff
@@ -18,41 +16,19 @@
...
-  checkboxes.on('change', function() {
-    $(this).parents('tr').toggleClass('info');
-  });
-
-  $('#select-all').on('change', function() {
-    checkboxes.prop("checked", $(this).prop("checked"));
-    checkboxes.parents('tr').toggleClass('info');
-  });
-});
-</script>
```
Comment (AndrewVos) [inline app/views/users/index.html.erb]: YESSSS. Glad this is getting removed from the view <3

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (AndrewVos) [inline app/views/layouts/application.html.erb]: Yep! Maybe store if the module has been loaded on the corresponding div like `data-loaded-modules='whatevs'`.

**Comment by Guntrisoft** [comment]
@AndrewVos happy enough for this to be merged now?

**Comment by AndrewVos** [comment]
@Guntrisoft lgtm

**Pull request merged**